pkg/networkservice/common/crossnse/server.go

```diff
@@ -0,0 +1,172 @@
+//				one by one and if request came back from cross nse, it will connect to a proper next client endpoint.
+//             - server - *registry.NetworkServiceRegistryServer.  Since registry.NetworkServiceRegistryServer is an interface
+//                        (and thus a pointer) *registry.NetworkServiceRegistryServer is a double pointer.  Meaning it
+//                        points to a place that points to a place that implements registry.NetworkServiceRegistryServer
+//                        This is done so that we can return a registry.NetworkServiceRegistryServer chain element
+//                        while maintaining the NewServer pattern for use like anything else in a chain.
+//                        The value in *server must be included in the registry.NetworkServiceRegistryServer listening
+//                        so it can capture the registrations.
+func NewServer(registryServer *registry.NetworkServiceEndpointRegistryServer) networkservice.NetworkServiceServer {
+	rv := &crossConnectNSEServer{}
+	*registryServer = crossnse.NewNetworkServiceRegistryServer(rv)
+	return rv
+}
+
+func (l *crossConnectNSEServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (result *networkservice.Connection, err error) {
+	// Check if there is no active connection, we need to replace endpoint url with forwarder url
+	conn := request.GetConnection()
+	ind := conn.GetPath().GetIndex() // It is designed to be used inside Endpoint, so current index is Endpoint already
+	connID := conn.GetId()
+
+	if len(conn.GetPath().GetPathSegments()) == 0 || ind <= 0 {
+		return nil, errors.Errorf("path segment doesn't have a client or cross connect nse identity")
+	}
+	// Remember current segment Name if not yet.
+	l.nsmgrInit.Do(func() {
+		l.nsmgrName = conn.GetPath().GetPathSegments()[ind].Name
+	})
+
+	// We need to find an Id from path to match active connection request.
+	clientConnID := l.getConnectionID(conn)
+
+	connInfoRaw, ok := l.activeConnection.Load(clientConnID)
+	if !ok {
+		if connID != clientConnID {
+			return nil, errors.Errorf("connection id should match current path segment id")
+		}
+		// We came from client, so select cross nse and go to it.
+		clientURL := clienturl.ClientURL(ctx)
+
+		// Iterate over all cross connect NSEs to check one with passed state.
+
+		l.crossNSEs.Range(func(key, value interface{}) bool {
+			crossNSE, ok := value.(*registry.NetworkServiceEndpoint)
+			if ok {
+				crossNSEURL, _ := url.Parse(crossNSE.Url)
+				crossCTX := clienturl.WithClientURL(ctx, crossNSEURL)
+
+				// Store client connection and selected cross connection URL.
+				_, _ = l.activeConnection.LoadOrStore(conn.Id, &connectionInfo{
+					endpointURL: clientURL,
+					crossNSEURL: crossNSEURL,
+				})
+				result, err = next.Server(crossCTX).Request(crossCTX, request)
+				if err != nil {
+					trace.Log(ctx).Errorf("failed to request cross NSE %v err: %v", crossNSEURL, err)
```

Comment:
Do we need to call `next.Server(crossCTX).Close()`?

Reply:
Yes, we need, Thanks.

Reply:
Nope, it is not required.